tests/error-qvm-tests.lisp

```diff
@@ -0,0 +1,202 @@
+      (assert (null decls))
+      (let ((confidence-window (sqrt (/ (log (/ 2 (- 1 alpha))) (* 2 trials)))))
+        `(deftest ,name ()
+           ,@(when docstring `(,docstring))
+           (let ((,histogram (make-hash-table)))
+             (dotimes (j ,trials)
+               (let* ((quil-string (format nil "DECLARE ro BIT[~A]" ,num-qubits))
+                      (memory-model (qvm:memory-descriptors-to-qvm-memory-model
+                                     (quil:parsed-program-memory-definitions
+                                      (quil::parse-quil quil-string))))
+                      (qvm (qvm.error::make-error-qvm ,num-qubits
+                                                      :classical-memory-model memory-model)))
+                 ;; enact clauses
+                 ,@(loop :for (setf-form . transitions) :in clauses
+                         :collect `(progn
+                                     ,setf-form
+                                     ,@(loop :for transition :in transitions
+                                             :collect `(qvm::transition qvm ,transition))))
+                 ;; record results
+                 (qvm::dereference-mref qvm (cl-quil::mref "ro" 0))
+                 (loop :with acc := 0
+                       :for i :from (1- ,num-qubits) :downto 0
+                       :do (setf acc (* acc 2)
+                                 acc (+ acc (qvm::dereference-mref qvm (cl-quil::mref "ro" i))))
+                       :finally (incf (gethash acc ,histogram 0)))))
+             ;; check our estimates
+             (maphash (lambda (key true-value)
+                        (let* ((recorded-hits (gethash key ,histogram 0d0))
+                               (estimated-value (/ recorded-hits ,trials)))
+                          (format *error-output* "Checking ~8b: ~5f < ~5f < ~5f, estimated mean: ~d / ~d = ~5f~%"
+                                  key
+                                  (- estimated-value ,confidence-window)
+                                  true-value
+                                  (+ estimated-value ,confidence-window)
+                                  (gethash key ,histogram 0d0)
+                                  ,trials
+                                  estimated-value)))
+                      (alexandria:plist-hash-table ,distribution))
+             (maphash (lambda (key true-value)
+                        (let* ((recorded-hits (gethash key ,histogram 0d0))
+                               (estimated-value (/ recorded-hits ,trials)))
+                          (is (< (- estimated-value ,confidence-window)
+                                 true-value
+                                 (+ estimated-value ,confidence-window)))))
+                      (alexandria:plist-hash-table ,distribution))))))))
+
+(define-depolarizing-test test-identity-noise
+    (:num-qubits 1
+     :distribution '(#b0 #.(- 1 (* 0.1 2/3)) #b1 #.(* 0.1 2/3))
```

Comment:
is it ok that these and elsewhere will be single floats? seems fine but I'm not sure

Reply:
🤷‍♀️ no harm changing them